src/src/com/microsoft/aad/adal/StorageHelper.java

```
@@ -401,69 +509,44 @@ final private SecretKey generateSecretKey() throws NoSuchAlgorithmException {
-        if (sSecretKeyFromAndroidKeyStore != null) {
-            return sSecretKeyFromAndroidKeyStore;
-        }
-
-        // Store secret key in a file after wrapping
-        File keyFile = new File(mContext.getDir(mContext.getPackageName(), Context.MODE_PRIVATE),
-                ADALKS);
-
-        if (mKeyPair == null) {
-            mKeyPair = getKeyPairFromAndroidKeyStore();
-            Logger.v(TAG, "Retrived keypair from androidKeyStore");
-        }
-
-        Cipher wrapCipher = Cipher.getInstance(WRAP_ALGORITHM);
-        // If keyfile does not exist, it needs to generate one
+    private synchronized SecretKey getUnwrappedSecretKey(final File keyFile) throws GeneralSecurityException, IOException {
+        Logger.v(TAG, "Reading SecretKey");
+        SecretKey unwrappedSecretKey = null;
+        
         if (!keyFile.exists()) {
-            Logger.v(TAG, "Key file does not exists");
-            final SecretKey key = generateSecretKey();
-            Logger.v(TAG, "Wrapping SecretKey");
-            final byte[] keyWrapped = wrap(wrapCipher, key);
-            Logger.v(TAG, "Writing SecretKey");
-            writeKeyData(keyFile, keyWrapped);
-            Logger.v(TAG, "Finished writing SecretKey");
+            throw new IOException("Key file to read does not exist");
         }
-
-        // Read from file again
-        Logger.v(TAG, "Reading SecretKey");
+        
         try {
-            final byte[] encryptedKey = readKeyData(keyFile);
-            if (encryptedKey == null || encryptedKey.length == 0) {
+            final byte[] wrappedSecretKey = readKeyData(keyFile);
+            if (wrappedSecretKey == null || wrappedSecretKey.length == 0) {
                 throw new UnrecoverableKeyException("Couldn't find encrypted key in file");
             }
             
-            sSecretKeyFromAndroidKeyStore = unwrap(wrapCipher, encryptedKey);
+            if (privateKey == null) {
```

1) I don't see initialization of privateKey
2) I don't think that this error should be recoverable, if this happened it means something went wrong and we swallowed some keystore exception. Do not make it recoverable